### PR TITLE
dispatches to echo

### PR DESCRIPTION
**Pull request by zylava**
@RyanLPeterman @adiraju13 Could one of you review my code?
Currently, it dispatches to echo correctly, dispatches to NotFound if prefix not found correctly, does not handle static file serving.
We still need to talk about the design of dispatching to StaticHandler.

### CODE REVIEW

File: handler.cpp [hunk 1]
```diff
@@ -25,18 +38,38 @@ RequestHandler::Status StaticHandler::Init(const std::string& uri_prefix, const
...
+    response->AddHeader("Content-Length", std::to_string(to_send.size()));
+    response->AddHeader("Content-Type", content_type);
+    std::string to_send_string(to_send.begin(), to_send.end());
+    response->SetBody(to_send_string);*/
     return RequestHandler::PASS;
 }
 
 std::string StaticHandler::get_path_from_url(std::string url) {
   // Assumption: url must be prefixed with "/static/" at this point
   //             get_function_from_url has already been called on url
```
Comment (adiraju13) [inline handler.cpp:68]: What is this function supposed to do?

Comment (adiraju13) [reply handler.cpp:68]: Like why would we use? could you add a comment to clarify that?

**Review by adiraju13 — COMMENTED** [review]
Please clarify about the code I commented on (its use, what the function does, etc).

File: handler.cpp [hunk 2]
```diff
@@ -7,11 +7,24 @@
...
+RequestHandler::Status NotFoundHandler::Init(const std::string& uri_prefix, const NginxConfig& config) {
+    return RequestHandler::PASS;
+}
+
+RequestHandler::Status NotFoundHandler::HandleRequest(const Request& request, Response* response){
+    response->SetStatus(Response::NOT_FOUND);
+    response->AddHeader("Content-Length", std::to_string(request.raw_request().size() - 4));
+    response->AddHeader("Content-Type", "text/plain");
+    response->SetBody("404 NOT FOUND");
+    return RequestHandler::PASS; //should this be FAIL actually?
```
Comment (adiraju13) [inline handler.cpp:19]: This is good to pass because it does its job successfully I think

File: handler.cpp [hunk 3]
```diff
@@ -25,18 +38,38 @@ RequestHandler::Status StaticHandler::Init(const std::string& uri_prefix, const
     return RequestHandler::PASS;
 }
 
-// TODO
+// TODO: url is empty now, might need to change Init? or somehow modify config to have url in it. Also for files that cannot be found, NotFoundHandler has to be invoked. How do we do that?
 RequestHandler::Status StaticHandler::HandleRequest(const Request& request, Response* response) {
+    /*std::cout << "========= Handling Static =========" << std::endl;
+    std::string abs_path = get_exec_path() + "/public" + get_path_from_url(url);
+    std::cout << "Serving file from: " << abs_path << std::endl;
```
Comment (adiraju13) [inline handler.cpp:45]: I do not think this code is valid anymore, but I will take care of static handling so don't worry about it. Just wanted to point that out now.

File: session.cpp [hunk 4]
```diff
@@ -97,10 +104,29 @@ std::string session::convert_buffer()
 }
 
 
-
-std::string session::get_function_from_url(std::string url)
+//get the longest prefix that matches with what's specified in config. If no match is found, an empty string is returned
+std::string session::get_function_from_url(const std::string original_url)
 {
+  std::string url = original_url + "/"; //so that we can handle the part after the last slash
```
Comment (adiraju13) [inline session.cpp:110]: Maybe check if the string already ends with a slash first before adding one?

File: session.cpp [hunk 5]
```diff
@@ -97,10 +104,29 @@ std::string session::convert_buffer()
...
   std::string function = "";
+  int startPos = url.find("/", 0);
+  int upto = url.length();
+  int lastPos;
+  std::string subUrl;
+  while (true){
+    lastPos = url.rfind("/", upto);
+    if (lastPos == 0){
```
Comment (adiraju13) [inline session.cpp:118]: if the string is just "/" will this break? Won't it break out of the loop and just return ""?

**Review by adiraju13 — COMMENTED** [review]
I think there might be corner case errors associated with get_function_from_url and the use of rfind. Check out the comments and let me know.

File: session.cpp (hunk 5 above)
Comment (zylava) [reply session.cpp:118]: Yes, good catch! I will close this pull request and submit another pull request.

File: session.cpp (hunk 4 above)
Comment (zylava) [reply session.cpp:110]: This will be handled in the pull request I am about to submit too.

**Pull request closed without merging**